snmp_exporter/collector.py

```diff
@@ -5,21 +5,42 @@
 
 from prometheus_client import Metric, CollectorRegistry, generate_latest, Gauge
 
-def walk_oids(host, port, oids, community):
-  session = netsnmp.Session(Version=2, DestHost=host, RemotePort=port,
-      Community=community, UseNumeric=True, Retries=3)
+def walk_oids(host, port, oids, version, auth):
+  if version == 3:
+    auth_user = auth['username']
+    auth_pass = auth['password']
+    auth_protocol = auth.get('auth_protocol', 'MD5')
+    priv_protocol = auth.get('priv_protocol', 'DES')
+    priv_pass = auth['priv_password']
+    sec_level = auth.get('security_level', 'noAuthNoPriv')
+    session = netsnmp.Session(Version=3, DestHost=host, RemotePort=port,
+        SecName=auth_user, AuthPass=auth_pass, AuthProto=auth_protocol,
+        PrivProto=priv_protocol, PrivPass=priv_pass, SecLevel=sec_level,
+        UseNumeric=True, Retries=3)
+  elif version == 2:
+    community = auth.get('community', 'public')
+    session = netsnmp.Session(Version=2, DestHost=host, RemotePort=port,
+        Community=community, UseNumeric=True, Retries=3)
+  elif version == 1:
+    community = auth.get('community', 'public')
+    session = netsnmp.Session(Version=1, DestHost=host, RemotePort=port,
+        Community=community, UseNumeric=True, Retries=3)
   for oid in oids:
-    for v in walk_oid(session, oid):
+    for v in walk_oid(session, version, oid):
       yield v
 
-def walk_oid(session, oid):
+def walk_oid(session, version, oid):
     last_oid = oid
     while True:
       # getbulk starts from the last oid we saw.
       vl = netsnmp.VarList(netsnmp.Varbind('.' + last_oid))
-      if not session.getbulk(0, 25, vl):
-        return
-
+      if version == 1:
+        if not session.getnext(vl):
```

Comment:
It seems that if the index of table being iterated isn't the penultimate item in the oid, this function will just keep iterating in an infinite loop.

My preferred fix for this is just to use session.walk - all the back and forth is then handled in net-snmp c and I believe it mimics the getbulk most closely. For example:

```python
if not session.walk(vl):
  return
```

Alternatively you could check you're not in an infinite loop like so

```python
next_oid = v.tag[1:] + '.' + v.iid
if not (last_oid + '.').startswith(oid + '.') or next_oid == last_oid:
  return
last_oid = next_oid
```